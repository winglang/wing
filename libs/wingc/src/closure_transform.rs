use indexmap::IndexMap;

use crate::{
	ast::{
		AccessModifier, ArgList, AssignmentKind, CalleeKind, Class, ClassField, Expr, ExprKind, FunctionBody,
		FunctionDefinition, FunctionParameter, FunctionSignature, Literal, New, Phase, Reference, Scope, Stmt, StmtKind,
		Symbol, TypeAnnotation, TypeAnnotationKind, UserDefinedType,
	},
	diagnostic::WingSpan,
	fold::{self, Fold},
	type_check::{CLASS_INFLIGHT_INIT_NAME, CLASS_INIT_NAME, CLOSURE_CLASS_HANDLE_METHOD},
};

pub const CLOSURE_CLASS_PREFIX: &str = "$Closure";
pub const PARENT_THIS_NAME: &str = "__parent_this";

/// Transforms inflight closures defined in preflight scopes into preflight classes.
///
/// This is done by wrapping the closure's code in a preflight class with a single method,
/// and replacing the closure with a reference to an instance of that class.
/// The class is given a unique name to avoid collisions.
///
/// For example, the following code:
///
/// ```wing
/// let b = new cloud.Bucket();
/// let f = inflight (message: str) => {
///   b.put("file.txt", message);
/// };
/// ```
///
/// is transformed into:
///
/// ```wing
/// let b = new cloud.Bucket();
/// class $Closure1 {
///   new() {}
///   inflight handle(message: str) {
///     b.put("file.txt", message);
///   }
/// }
/// let f = new $Closure1();
/// ```
pub struct ClosureTransformer {
	// Whether the transformer is inside a preflight or inflight scope.
	// Only inflight closures defined in preflight scopes need to be transformed.
	phase: Phase,
	// Whether the transformer is inside a scope where "this" is valid.
	inside_scope_with_this: bool,
	// Helper state for generating unique class names
	closure_counter: usize,
	// Stores the list of class definitions that need to be added to the nearest scope
	class_statements: Vec<Stmt>,
	// Track the statement index of the nearest statement we're inside so that
	// newly-inserted classes will have access to the correct state
	nearest_stmt_idx: usize,
}

impl ClosureTransformer {
	pub fn new() -> Self {
		Self {
			phase: Phase::Preflight,
			inside_scope_with_this: false,
			closure_counter: 0,
			class_statements: vec![],
			nearest_stmt_idx: 0,
		}
	}
}

impl Fold for ClosureTransformer {
	fn fold_scope(&mut self, node: Scope) -> Scope {
		let mut statements = vec![];

		for stmt in node.statements {
			// TODO: can we remove "idx" from Stmt to avoid having to reason about this?
			// or add a compiler step that updates statement indices after folding?
			let old_nearest_stmt_idx = self.nearest_stmt_idx;
			self.nearest_stmt_idx = stmt.idx;
			let new_stmt = self.fold_stmt(stmt);
			self.nearest_stmt_idx = old_nearest_stmt_idx;

			// Add any new statements we have accumulated that define temporary classes
			statements.append(&mut self.class_statements);

			// Push the folded statement last. This way it has access to all of the temporary
			// classes we have defined.
			statements.push(new_stmt);

			// Reset scope-specific state
			self.class_statements.clear();
		}

		Scope {
			id: node.id,
			statements,
			span: node.span,
		}
	}

	fn fold_function_definition(&mut self, node: FunctionDefinition) -> FunctionDefinition {
		let prev_phase = self.phase;
		self.phase = node.signature.phase;
		let prev_inside_scope_with_this = self.inside_scope_with_this;
		if !node.is_static {
			self.inside_scope_with_this = true;
		}
		let new_node = fold::fold_function_definition(self, node);
		self.inside_scope_with_this = prev_inside_scope_with_this;
		self.phase = prev_phase;
		new_node
	}

	fn fold_expr(&mut self, expr: Expr) -> Expr {
		// Inflight closures that are themselves defined inflight do not need
		// to be transformed, since they are not created in preflight.
		if self.phase == Phase::Inflight {
			return expr;
		}

		// If we encounter a preflight closure, we don't need to
		// transform it into a preflight class, but we still want to recurse
		// in case its body contains any inflight closures.
		if let ExprKind::FunctionClosure(ref func_def) = expr.kind {
			if func_def.signature.phase != Phase::Inflight {
				return fold::fold_expr(self, expr);
			}
		}

		match expr.kind {
			ExprKind::FunctionClosure(func_def) => {
				self.closure_counter += 1;

				let file_id = &expr.span.file_id;

				let new_class_name = Symbol {
					name: format!("{}{}", CLOSURE_CLASS_PREFIX, self.closure_counter),
					span: WingSpan::for_file(file_id),
				};
				let handle_name = Symbol {
					name: CLOSURE_CLASS_HANDLE_METHOD.to_string(),
					span: WingSpan::for_file(file_id),
				};

				let class_udt = UserDefinedType {
					root: new_class_name.clone(),
					fields: vec![],
					span: WingSpan::for_file(file_id),
				};

				let class_type_annotation = TypeAnnotation {
					kind: TypeAnnotationKind::UserDefined(class_udt.clone()),
					span: WingSpan::for_file(file_id),
				};

				let class_fields: Vec<ClassField> = vec![];
				let class_init_params: Vec<FunctionParameter> = vec![];

				let new_func_def = if self.inside_scope_with_this {
					// If we are inside a class, we transform inflight closures with an extra
					// `let __parent_this_${CLOSURE_COUNT} = this;` statement before the class definition, and replace references
					// to `this` with `__parent_this_${CLOSURE_COUNT}` so that they can access the parent class's fields.
					let parent_this = format!("{}_{}", PARENT_THIS_NAME, self.closure_counter);
					let mut this_transform = RenameThisTransformer {
						from: "this",
						to: parent_this.as_str(),
						inside_class: false,
					};
					this_transform.fold_function_definition(func_def)
				} else {
					func_def
				};

				let new_func_def = FunctionDefinition {
					name: Some(handle_name.clone()),
					body: new_func_def.body,
					signature: new_func_def.signature,
					span: new_func_def.span,

					// Anonymous functions are always static -- since the function code is now an instance method on a class,
					// we need to set this to false.
					is_static: false,
					access: AccessModifier::Public,
				};

				// class_init_body :=
				// ```
				// nodeof(this).hidden = true;
				// ```
				let std_display_of_this = Expr::new(
					ExprKind::Call {
						callee: CalleeKind::Expr(Box::new(Expr::new(
							ExprKind::Reference(Reference::Identifier(Symbol::new(
								"nodeof",
								WingSpan::for_file(file_id),
							))),
							WingSpan::for_file(file_id),
						))),
						arg_list: ArgList {
							named_args: IndexMap::new(),
							pos_args: vec![Expr::new(
								ExprKind::Reference(Reference::Identifier(Symbol::new("this", WingSpan::for_file(file_id)))),
								WingSpan::for_file(file_id),
							)],
							span: WingSpan::for_file(file_id),
						},
					},
					WingSpan::for_file(file_id),
				);
				let class_init_body = vec![Stmt {
					idx: 0,
					kind: StmtKind::Assignment {
						kind: AssignmentKind::Assign,
						variable: Reference::InstanceMember {
							object: Box::new(std_display_of_this),
							property: Symbol::new("hidden", WingSpan::for_file(file_id)),
							optional_accessor: false,
						},

						value: Expr::new(ExprKind::Literal(Literal::Boolean(true)), WingSpan::for_file(file_id)),
					},
					span: WingSpan::for_file(file_id),
				}];

				// If we are inside a scope with "this", add define `let __parent_this_${CLOSURE_COUNT} = this` which can be
				// used by the newly-created preflight classes
				if self.inside_scope_with_this {
					let parent_this_name = Symbol::new(
						format!("{}_{}", PARENT_THIS_NAME, self.closure_counter),
						WingSpan::for_file(file_id),
					);
					let this_name = Symbol::new("this", WingSpan::for_file(file_id));
					let parent_this_def = Stmt {
						kind: StmtKind::Let {
							reassignable: false,
							var_name: parent_this_name,
							initial_value: Expr::new(
								ExprKind::Reference(Reference::Identifier(this_name)),
								WingSpan::for_file(file_id),
							),
							type_: None,
						},
						span: WingSpan::for_file(file_id),
						idx: 0,
					};
					self.class_statements.push(parent_this_def);
				}

				// class_def :=
				// ```
				// class <new_class_name> {
				//   new(<class_init_params>) {<class_init_body>}
				//   inflight handle() {
				//     <transformed_def>
				//   }
				// }
				// ```
				let class_def = Stmt {
					kind: StmtKind::Class(Class {
						name: new_class_name.clone(),
						span: new_func_def.span.clone(),
						phase: Phase::Preflight,
						initializer: FunctionDefinition {
							name: Some(CLASS_INIT_NAME.into()),
							signature: FunctionSignature {
								parameters: class_init_params,
								return_type: Box::new(class_type_annotation.clone()),
								phase: Phase::Preflight,
							},
							is_static: true,
							body: FunctionBody::Statements(Scope::new(class_init_body, WingSpan::for_file(file_id))),
							span: WingSpan::for_file(file_id),
							access: AccessModifier::Public,
						},
						fields: class_fields,
						implements: vec![],
						parent: None,
						methods: vec![(handle_name.clone(), new_func_def)],
						inflight_initializer: FunctionDefinition {
							name: Some(CLASS_INFLIGHT_INIT_NAME.into()),
							signature: FunctionSignature {
								parameters: vec![],
								return_type: Box::new(TypeAnnotation {
									kind: TypeAnnotationKind::Void,
									span: WingSpan::for_file(file_id),
								}),
								phase: Phase::Inflight,
							},
							is_static: false,
							body: FunctionBody::Statements(Scope::new(vec![], WingSpan::for_file(file_id))),
							span: WingSpan::for_file(file_id),
							access: AccessModifier::Public,
						},
						access: AccessModifier::Private,
						auto_id: true,
					}),
					idx: self.nearest_stmt_idx,
					span: WingSpan::for_file(file_id),
				};

				// new_class_instance :=
				// ```
				// new <new_class_name>();
				// ```
				let new_class_instance = Expr::new(
					ExprKind::New(New {
						class: class_udt,
						arg_list: ArgList {
							named_args: IndexMap::new(),
							pos_args: vec![],
							span: WingSpan::for_file(file_id),
						},
						obj_id: None,
						obj_scope: None,
					}),
					expr.span.clone(), // <<-- span of original expression
				);

				self.class_statements.push(class_def);

				new_class_instance
			}
			_ => fold::fold_expr(self, expr),
		}
	}
}

/// Rename all occurrences of an identifier inside an inflight closure to a new name.
struct RenameThisTransformer<'a> {
	from: &'a str,
	to: &'a str,
	// The transform shouldn't change references to `this` inside inflight classes since
	// they refer to different objects.
	inside_class: bool,
}

impl Default for RenameThisTransformer<'_> {
	fn default() -> Self {
		Self {
			from: "this",
			to: PARENT_THIS_NAME,
			inside_class: false,
		}
	}
}

impl<'a> Fold for RenameThisTransformer<'a> {
	fn fold_class(&mut self, node: Class) -> Class {
		let old_inside_class = self.inside_class;
		self.inside_class = true;
		let new_class = fold::fold_class(self, node);
		self.inside_class = old_inside_class;
		new_class
	}

	fn fold_reference(&mut self, node: Reference) -> Reference {
		match node {
			Reference::Identifier(ident) => {
				if !self.inside_class && ident.name == self.from {
					Reference::Identifier(Symbol {
						name: self.to.to_string(),
						span: ident.span,
					})
				} else {
					Reference::Identifier(ident)
				}
			}
			Reference::InstanceMember { .. } | Reference::TypeMember { .. } | Reference::ElementAccess { .. } => {
				fold::fold_reference(self, node)
			}
		}
	}
}
