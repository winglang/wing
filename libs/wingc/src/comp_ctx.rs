use std::{
	backtrace::{Backtrace, BacktraceStatus},
	cell::RefCell,
};

use strum::{Display, EnumString};

use crate::diagnostic::{report_diagnostic, Diagnostic, WingSpan};

/// The different phases of compilation, used for tracking compilation context
/// for diagnostic purposes. Feel free to add new phases as needed.
/// Try to make these end with 'ing' (building, parsing, etc.) so they'll fit in
/// in the context of a diagnostic message.
/// See `CompilationContext::set` for more information.
#[derive(EnumString, Display, PartialEq, Clone, Copy, Debug, Default)]
#[strum(serialize_all = "kebab-case")]
pub enum CompilationPhase {
	#[default]
	Compiling,
	Parsing,
	TypeChecking,
	Lifting,
	Jsifying,
}

pub struct CompilationContext {
	/// Description of current compilation phase
	pub phase: CompilationPhase,

	/// Location in source we're currently processing
	pub span: WingSpan,
}

thread_local! {
	pub static COMPILATION_CONTEXT: RefCell<CompilationContext> = RefCell::new(CompilationContext {
		phase: CompilationPhase::default(),
		span: WingSpan::default(),
	});
}

impl CompilationContext {
	/// Set global information about the current compilation phase.
	/// This is used for diagnostics, specifically for custom panic messages.
	///
	/// # Arguments
	///
	/// * `phase` - Description of current compilation phase.
	/// * `span` - Location in source we're currently processing.
	pub fn set(phase: CompilationPhase, span: &WingSpan) {
		COMPILATION_CONTEXT.with(|c| {
			c.replace_with(|_| CompilationContext {
				phase,
				span: span.clone(),
			});
		});
	}

	pub fn get_phase() -> CompilationPhase {
		COMPILATION_CONTEXT.with(|c| c.borrow().phase)
	}

	fn get_span() -> WingSpan {
		COMPILATION_CONTEXT.with(|c| c.borrow().span.clone())
	}
}

/// Macro used for explicit panics if the environment variable `WINGC_DEBUG_PANIC` is set.
/// This can be used if we want to conditionally panic in certain situations.
/// This is a macro and not a function so we can get the location of the caller
/// in the panic message.
#[macro_export]
macro_rules! dbg_panic {
	() => {{
		|| -> () {
			// Get environment variable to see if we should panic or not
			let Ok(dbg_panic) = std::env::var("WINGC_DEBUG_PANIC") else {
				return;
			};

			if dbg_panic == "1"
				|| dbg_panic == "true"
				|| (dbg_panic
					.parse::<$crate::comp_ctx::CompilationPhase>()
					.map(|p| p == $crate::comp_ctx::CompilationContext::get_phase())
					.unwrap_or(false))
			{
				panic!("User invoked panic");
			}
		}();
	}};
}

pub fn set_custom_panic_hook() {
	std::panic::set_hook(Box::new(|pi| {
		// Print backtrace if RUST_BACKTRACE=1
		let bt = Backtrace::capture();
		if bt.status() == BacktraceStatus::Captured {
			eprintln!("Panic backtrace:\n{}", bt);
		} else {
			eprintln!("Panicked, backtrace not captured: {:?}", bt.status());
		}

		report_diagnostic(Diagnostic {
			message: format!(
				"Compiler bug ('{}' at {}) during {}, please report at https://www.winglang.io/contributing/start-here/bugs",
				pi.payload().downcast_ref::<&str>().unwrap_or(&""),
				pi.location().unwrap_or(&std::panic::Location::caller()),
				CompilationContext::get_phase()
			),
			span: Some(CompilationContext::get_span()),
			annotations: vec![],
		})
	}));
}
