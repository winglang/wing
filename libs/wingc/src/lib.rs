#![allow(clippy::all)]
#![deny(clippy::correctness)]
#![deny(clippy::suspicious)]
#![deny(clippy::complexity)]
#![allow(clippy::vec_box)]

#[macro_use]
extern crate lazy_static;

use ast::{AccessModifier, Scope, Symbol, UtilityFunctions};
use camino::{Utf8Path, Utf8PathBuf};
use closure_transform::ClosureTransformer;
use comp_ctx::set_custom_panic_hook;
use diagnostic::{found_errors, report_diagnostic, Diagnostic};
use file_graph::FileGraph;
use files::Files;
use fold::Fold;
use indexmap::IndexMap;
use jsify::JSifier;

use lifting::LiftVisitor;
use parser::parse_wing_project;
use struct_schema::StructSchemaVisitor;
use type_check::jsii_importer::JsiiImportSpec;
use type_check::symbol_env::{StatementIdx, SymbolEnvKind};
use type_check::type_reference_transform::TypeReferenceTransformer;
use type_check::{FunctionSignature, SymbolKind, Type};
use type_check_assert::TypeCheckAssert;
use valid_json_visitor::ValidJsonVisitor;
use visit::Visit;
use wasm_util::{ptr_to_str, string_to_combined_ptr, WASM_RETURN_ERROR};
use wingii::type_system::TypeSystem;

use crate::docs::Docs;
use crate::parser::normalize_path;
use std::alloc::{alloc, dealloc, Layout};

use std::mem;

use crate::ast::Phase;
use crate::type_check::symbol_env::SymbolEnv;
use crate::type_check::{FunctionParameter, TypeChecker, Types};

#[macro_use]
#[cfg(test)]
mod test_utils;

pub mod ast;
pub mod closure_transform;
mod comp_ctx;
pub mod debug;
pub mod diagnostic;
mod docs;
mod file_graph;
mod files;
pub mod fold;
pub mod jsify;
pub mod json_schema_generator;
mod lifting;
pub mod lsp;
pub mod parser;
pub mod struct_schema;
pub mod type_check;
mod type_check_assert;
mod valid_json_visitor;
pub mod visit;
mod visit_context;
mod visit_types;
mod wasm_util;

const WINGSDK_ASSEMBLY_NAME: &'static str = "@winglang/sdk";

pub const WINGSDK_STD_MODULE: &'static str = "std";
const WINGSDK_CLOUD_MODULE: &'static str = "cloud";
const WINGSDK_UTIL_MODULE: &'static str = "util";
const WINGSDK_HTTP_MODULE: &'static str = "http";
const WINGSDK_MATH_MODULE: &'static str = "math";
const WINGSDK_AWS_MODULE: &'static str = "aws";
const WINGSDK_EX_MODULE: &'static str = "ex";
const WINGSDK_EXPECT_MODULE: &'static str = "expect";
const WINGSDK_REGEX_MODULE: &'static str = "regex";
const WINGSDK_FS_MODULE: &'static str = "fs";
const WINGSDK_SIM_MODULE: &'static str = "sim";
const WINGSDK_UI_MODULE: &'static str = "ui";

pub const UTIL_CLASS_NAME: &'static str = "Util";

const WINGSDK_BRINGABLE_MODULES: [&'static str; 11] = [
	WINGSDK_CLOUD_MODULE,
	WINGSDK_UTIL_MODULE,
	WINGSDK_HTTP_MODULE,
	WINGSDK_MATH_MODULE,
	WINGSDK_AWS_MODULE,
	WINGSDK_EX_MODULE,
	WINGSDK_EXPECT_MODULE,
	WINGSDK_REGEX_MODULE,
	WINGSDK_FS_MODULE,
	WINGSDK_SIM_MODULE,
	WINGSDK_UI_MODULE,
];

const WINGSDK_GENERIC: &'static str = "std.T1";
const WINGSDK_DURATION: &'static str = "std.Duration";
const WINGSDK_MAP: &'static str = "std.Map";
const WINGSDK_MUT_MAP: &'static str = "std.MutMap";
const WINGSDK_ARRAY: &'static str = "std.Array";
const WINGSDK_MUT_ARRAY: &'static str = "std.MutArray";
const WINGSDK_SET: &'static str = "std.Set";
const WINGSDK_MUT_SET: &'static str = "std.MutSet";
const WINGSDK_STRING: &'static str = "std.String";
const WINGSDK_JSON: &'static str = "std.Json";
const WINGSDK_MUT_JSON: &'static str = "std.MutJson";
const WINGSDK_RESOURCE: &'static str = "std.Resource";
const WINGSDK_STRUCT: &'static str = "std.Struct";
const WINGSDK_TEST_CLASS_NAME: &'static str = "Test";
// const WINGSDK_NODE: &'static str = "std.Node";

const CONSTRUCT_BASE_CLASS: &'static str = "constructs.Construct";
const CONSTRUCT_BASE_INTERFACE: &'static str = "constructs.IConstruct";

const MACRO_REPLACE_SELF: &'static str = "$self$";
const MACRO_REPLACE_ARGS: &'static str = "$args$";
const MACRO_REPLACE_ARGS_TEXT: &'static str = "$args_text$";

pub const TRUSTED_LIBRARY_NPM_NAMESPACE: &'static str = "@winglibs";

pub struct CompilerOutput {}

/// Exposes an allocation function to the WASM host
///
/// _This implementation is copied from wasm-bindgen_
#[no_mangle]
pub unsafe extern "C" fn wingc_malloc(size: usize) -> *mut u8 {
	let align = mem::align_of::<usize>();
	let layout = Layout::from_size_align(size, align).expect("Invalid layout");
	if layout.size() > 0 {
		let ptr = alloc(layout);
		if !ptr.is_null() {
			return ptr;
		} else {
			std::alloc::handle_alloc_error(layout);
		}
	} else {
		return align as *mut u8;
	}
}

/// Expose a deallocation function to the WASM host
///
/// _This implementation is copied from wasm-bindgen_
#[no_mangle]
pub unsafe extern "C" fn wingc_free(ptr: *mut u8, size: usize) {
	// This happens for zero-length slices, and in that case `ptr` is
	// likely bogus so don't actually send this to the system allocator
	if size == 0 {
		return;
	}
	let align = mem::align_of::<usize>();
	let layout = Layout::from_size_align_unchecked(size, align);
	dealloc(ptr, layout);
}

/// Expose one time-initiliazation function to the WASM host,
/// should be called before any other function
#[no_mangle]
pub unsafe extern "C" fn wingc_init() {
	// Setup a custom panic hook to report panics as compilation diagnostics
	set_custom_panic_hook();
}

#[no_mangle]
pub unsafe extern "C" fn wingc_compile(ptr: u32, len: u32) -> u64 {
	let args = ptr_to_str(ptr, len);

	let split = args.split(";").collect::<Vec<&str>>();
	if split.len() != 3 {
		report_diagnostic(Diagnostic {
			message: format!("Expected 3 arguments to wingc_compile, got {}", split.len()),
			span: None,
			annotations: vec![],
			hints: vec![],
		});
		return WASM_RETURN_ERROR;
	}
	let source_path = Utf8Path::new(split[0]);
	let output_dir = split.get(1).map(|s| Utf8Path::new(s)).expect("output dir not provided");
	let project_dir = split
		.get(2)
		.map(|s| Utf8Path::new(s))
		.expect("project dir not provided");

	if !source_path.exists() {
		report_diagnostic(Diagnostic {
			message: format!("Source path cannot be found: {}", source_path),
			span: None,
			annotations: vec![],
			hints: vec![],
		});
		return WASM_RETURN_ERROR;
	}

	let results = compile(project_dir, source_path, None, output_dir);
	if results.is_err() {
		WASM_RETURN_ERROR
	} else {
		string_to_combined_ptr("".to_string())
	}
}

pub fn type_check(
	scope: &mut Scope,
	types: &mut Types,
	file_path: &Utf8Path,
	file_graph: &FileGraph,
	jsii_types: &mut TypeSystem,
	jsii_imports: &mut Vec<JsiiImportSpec>,
) {
	let mut env = types.add_symbol_env(SymbolEnv::new(None, SymbolEnvKind::Scope, Phase::Preflight, 0));
	types.set_scope_env(scope, env);

	add_builtin(
		UtilityFunctions::Log.to_string().as_str(),
		Type::Function(FunctionSignature {
			this_type: None,
			parameters: vec![FunctionParameter {
				name: "message".into(),
				typeref: types.string(),
				docs: Docs::with_summary("The message to log"),
				variadic: false,
			}],
			return_type: types.void(),
			phase: Phase::Independent,
			js_override: Some("console.log($args$)".to_string()),
			docs: Docs::with_summary("Logs a message"),
		}),
		scope,
		types,
	);
	add_builtin(
		UtilityFunctions::Assert.to_string().as_str(),
		Type::Function(FunctionSignature {
			this_type: None,
			parameters: vec![FunctionParameter {
				name: "condition".into(),
				typeref: types.bool(),
				docs: Docs::with_summary("The condition to assert"),
				variadic: false,
			}],
			return_type: types.void(),
			phase: Phase::Independent,
			js_override: Some("$helpers.assert($args$, \"$args_text$\")".to_string()),
			docs: Docs::with_summary("Asserts that a condition is true"),
		}),
		scope,
		types,
	);
	add_builtin(
		UtilityFunctions::UnsafeCast.to_string().as_str(),
		Type::Function(FunctionSignature {
			this_type: None,
			parameters: vec![FunctionParameter {
				name: "value".into(),
				typeref: types.anything(),
				docs: Docs::with_summary("The value to cast into a different type"),
				variadic: false,
			}],
			return_type: types.anything(),
			phase: Phase::Independent,
			js_override: Some("$args$".to_string()),
			docs: Docs::with_summary("Casts a value into a different type. This is unsafe and can cause runtime errors"),
		}),
		scope,
		types,
	);

	let mut tc = TypeChecker::new(types, file_path, file_graph, jsii_types, jsii_imports);
	tc.add_jsii_module_to_env(
		&mut env,
		WINGSDK_ASSEMBLY_NAME.to_string(),
		vec![WINGSDK_STD_MODULE.to_string()],
		&Symbol::global(WINGSDK_STD_MODULE),
		None,
	);

	tc.type_check_file_or_dir(file_path, scope);
}

// TODO: refactor this (why is scope needed?) (move to separate module?)
fn add_builtin(name: &str, typ: Type, scope: &mut Scope, types: &mut Types) {
	let sym = Symbol::global(name);
	let mut scope_env = types.get_scope_env(&scope);
	scope_env
		.define(
			&sym,
			SymbolKind::make_free_variable(sym.clone(), types.add_type(typ), false, Phase::Independent),
			AccessModifier::Private,
			StatementIdx::Top,
		)
		.expect("Failed to add builtin");
}

pub fn compile(
	project_dir: &Utf8Path,
	source_path: &Utf8Path,
	source_text: Option<String>,
	out_dir: &Utf8Path,
) -> Result<CompilerOutput, ()> {
	let source_path = normalize_path(source_path, None);

	// -- PARSING PHASE --
	let mut files = Files::new();
	let mut file_graph = FileGraph::default();
	let mut tree_sitter_trees = IndexMap::new();
	let mut asts = IndexMap::new();
	let topo_sorted_files = parse_wing_project(
		&source_path,
		source_text,
		&mut files,
		&mut file_graph,
		&mut tree_sitter_trees,
		&mut asts,
	);

	// -- DESUGARING PHASE --

	// Transform all inflight closures defined in preflight into single-method resources
	let mut asts = asts
		.into_iter()
		.map(|(path, scope)| {
			let mut inflight_transformer = ClosureTransformer::new();
			let scope = inflight_transformer.fold_scope(scope);
			(path, scope)
		})
		.collect::<IndexMap<Utf8PathBuf, Scope>>();

	// -- TYPECHECKING PHASE --

	// Create universal types collection (need to keep this alive during entire compilation)
	let mut types = Types::new();
	let mut jsii_types = TypeSystem::new();

	// Create a universal JSII import spec (need to keep this alive during entire compilation)
	let mut jsii_imports = vec![];

	// Type check all files in topological order (start with files that don't bring any other
	// Wing files, then move on to files that depend on those, and repeat)
	for file in &topo_sorted_files {
		let mut scope = asts.remove(file).expect("matching AST not found");
		type_check(
			&mut scope,
			&mut types,
			&file,
			&file_graph,
			&mut jsii_types,
			&mut jsii_imports,
		);

		// Make sure all type reference are no longer considered references
		let mut tr_transformer = TypeReferenceTransformer { types: &mut types };
		let scope = tr_transformer.fold_scope(scope);

		// Validate the type checker didn't miss anything - see `TypeCheckAssert` for details
		let mut tc_assert = TypeCheckAssert::new(&types, found_errors());
		tc_assert.check(&scope);

		// Validate all Json literals to make sure their values are legal
		let mut json_checker = ValidJsonVisitor::new(&types);
		json_checker.check(&scope);

		asts.insert(file.clone(), scope);
	}

	// Verify that the project dir is absolute
	if !is_absolute_path(&project_dir) {
		report_diagnostic(Diagnostic {
			message: format!("Project directory must be absolute: {}", project_dir),
			span: None,
			annotations: vec![],
			hints: vec![],
		});
		return Err(());
	}

	let mut jsifier = JSifier::new(&mut types, &files, &file_graph, &source_path);

	// -- LIFTING PHASE --

	let mut asts = asts
		.into_iter()
		.map(|(path, scope)| {
			let mut lift = LiftVisitor::new(&jsifier);
			lift.visit_scope(&scope);
			(path, scope)
		})
		.collect::<IndexMap<Utf8PathBuf, Scope>>();

	// bail out now (before jsification) if there are errors (no point in jsifying)
	if found_errors() {
		return Err(());
	}

	// -- STRUCT SCHEMA GENERATION PHASE --
	// Need to do this before jsification so that we know what struct schemas need to be generated
	asts = asts
		.into_iter()
		.map(|(path, scope)| {
			let mut reference_visitor = StructSchemaVisitor::new(&jsifier);
			reference_visitor.visit_scope(&scope);
			(path, scope)
		})
		.collect::<IndexMap<Utf8PathBuf, Scope>>();

	// -- JSIFICATION PHASE --

	for file in &topo_sorted_files {
		let scope = asts.get_mut(file).expect("matching AST not found");
		jsifier.jsify(file, &scope);
	}

	let files = jsifier.output_files.borrow_mut();
	match files.emit_files(out_dir) {
		Ok(()) => {}
		Err(err) => report_diagnostic(err.into()),
	}

	if found_errors() {
		return Err(());
	}

	return Ok(CompilerOutput {});
}

pub fn is_absolute_path(path: &Utf8Path) -> bool {
	if path.starts_with("/") {
		return true;
	}

	// Check if this is a Windows path instead by checking if the second char is a colon
	// Note: Cannot use Utf8Path::is_absolute() because it doesn't work with Windows paths on WASI
	let chars = path.as_str().chars().collect::<Vec<char>>();
	if chars.len() < 2 || chars[1] != ':' {
		return false;
	}

	return true;
}

#[cfg(test)]
mod sanity {
	use camino::{Utf8Path, Utf8PathBuf};

	use crate::{compile, diagnostic::assert_no_panics};
	use std::fs;

	fn get_wing_files<P>(dir: P) -> impl Iterator<Item = Utf8PathBuf>
	where
		P: AsRef<Utf8Path>,
	{
		fs::read_dir(dir.as_ref())
			.unwrap()
			.map(|entry| Utf8PathBuf::from_path_buf(entry.unwrap().path()).expect("invalid unicode path"))
			.filter(|path| path.is_file() && path.extension().map(|ext| ext == "w").unwrap_or(false))
	}

	fn compile_test(test_dir: &str, expect_failure: bool) {
		let test_dir = Utf8Path::new(test_dir).canonicalize_utf8().unwrap();
		for test_file in get_wing_files(&test_dir) {
			println!("\n=== {} ===\n", test_file);

			let out_dir = test_dir.join(format!("target/wingc/{}.out", test_file.file_name().unwrap()));

			// reset out_dir
			if out_dir.exists() {
				fs::remove_dir_all(&out_dir).expect("remove out dir");
			}

			let result = compile(&test_dir, &test_file, None, &out_dir);

			if result.is_err() {
				assert!(
					expect_failure,
					"{}: Expected compilation success, but failed: {:#?}",
					test_file,
					result.err().unwrap()
				);

				// Even if the test fails when we expect it to, none of the failures should be due to a compiler bug
				assert_no_panics();
			} else {
				assert!(
					!expect_failure,
					"{}: Expected compilation failure, but succeeded",
					test_file,
				);
			}
		}
	}

	#[test]
	fn can_compile_valid_files() {
		compile_test("../../examples/tests/valid", false);
	}

	#[test]
	fn can_compile_error_files() {
		compile_test("../../examples/tests/error", false);
	}

	#[test]
	fn cannot_compile_invalid_files() {
		compile_test("../../examples/tests/invalid", true);
	}
}
