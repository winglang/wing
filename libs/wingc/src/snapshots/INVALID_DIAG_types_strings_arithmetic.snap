---
source: libs/wingc/src/lib.rs
description: "../../examples/tests/invalid/types_strings_arithmetic.w"
---
[
    Diagnostic {
        message: "Expected type \"number\", but got \"string\" instead",
        span: Some(
            WingSpan {
                start: Point {
                    row: 0,
                    column: 15,
                },
                end: Point {
                    row: 0,
                    column: 18,
                },
                start_byte: 15,
                end_byte: 18,
                file_id: "../../examples/tests/invalid/types_strings_arithmetic.w",
            },
        ),
        level: Error,
    },
    Diagnostic {
        message: "Expected type \"number\", but got \"string\" instead",
        span: Some(
            WingSpan {
                start: Point {
                    row: 0,
                    column: 15,
                },
                end: Point {
                    row: 0,
                    column: 18,
                },
                start_byte: 15,
                end_byte: 18,
                file_id: "../../examples/tests/invalid/types_strings_arithmetic.w",
            },
        ),
        level: Error,
    },
    Diagnostic {
        message: "Expected type \"number\", but got \"string\" instead",
        span: Some(
            WingSpan {
                start: Point {
                    row: 0,
                    column: 15,
                },
                end: Point {
                    row: 0,
                    column: 18,
                },
                start_byte: 15,
                end_byte: 18,
                file_id: "../../examples/tests/invalid/types_strings_arithmetic.w",
            },
        ),
        level: Error,
    },
    Diagnostic {
        message: "Expected type \"number\", but got \"string\" instead",
        span: Some(
            WingSpan {
                start: Point {
                    row: 3,
                    column: 13,
                },
                end: Point {
                    row: 3,
                    column: 16,
                },
                start_byte: 85,
                end_byte: 88,
                file_id: "../../examples/tests/invalid/types_strings_arithmetic.w",
            },
        ),
        level: Error,
    },
    Diagnostic {
        message: "Expected type \"number\", but got \"string\" instead",
        span: Some(
            WingSpan {
                start: Point {
                    row: 9,
                    column: 29,
                },
                end: Point {
                    row: 9,
                    column: 40,
                },
                start_byte: 224,
                end_byte: 235,
                file_id: "../../examples/tests/invalid/types_strings_arithmetic.w",
            },
        ),
        level: Error,
    },
    Diagnostic {
        message: "Expected type \"number\", but got \"string\" instead",
        span: Some(
            WingSpan {
                start: Point {
                    row: 9,
                    column: 29,
                },
                end: Point {
                    row: 9,
                    column: 40,
                },
                start_byte: 224,
                end_byte: 235,
                file_id: "../../examples/tests/invalid/types_strings_arithmetic.w",
            },
        ),
        level: Error,
    },
]
